### PR TITLE
[Feature] Support jodatime_format function to partition pruning

### LINKED ISSUE #52680: Support jodatime_format function to partition pruning
- [x] https://github.com/StarRocks/starrocks/pull/52620

## Feature request

**Is your feature request related to a problem? Please describe.**

1.hive 外表testDb.testTable 分区字段是dt ，有6个分区
dt=20241101
dt=20241102
dt=20241103
dt=20241104
dt=20241105
dt=20241106
2.在执行下面SQL时，无法执行分区裁剪
select * from flight.zhimin_t where dt = jodatime_format('2024-11-02 15:58:51', 'yyyyMMdd')
**Describe the solution you'd like**

增加jodatime_format转换方法
**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by renzhimin7**
## Why I'm doing:
1. hive 外表testDb.testTable  分区字段是dt ，有6个分区
  dt=20241101
  dt=20241102
  dt=20241103
  dt=20241104
  dt=20241105
  dt=20241106
2. 在执行下面SQL时，无法执行分区裁剪
select * from flight.zhimin_t where dt = jodatime_format('2024-11-02 15:58:51', 'yyyyMMdd')
image

## What I'm doing:
增加jodatime_format转换方法
验证可以裁剪
image

Fixes #52680 

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
- [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java [hunk 1]
```diff
@@ -380,6 +380,28 @@ public static ConstantOperator dateFormat(ConstantOperator date, ConstantOperato
...
+        } else {
+            String result = date.getDatetime().format(DateTimeFormatter.ofPattern(fmtLiteral.getVarchar()));
+            return ConstantOperator.createVarchar(result);
+        }
+    }
+
+
     @ConstantFunction.List(list = {
             @ConstantFunction(name = "to_iso8601", argTypes = {DATETIME}, returnType = VARCHAR, isMonotonic = true),
             @ConstantFunction(name = "to_iso8601", argTypes = {DATE}, returnType = VARCHAR, isMonotonic = true)
```
[the remark this answers is not in the document]
Comment (renzhimin7) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java:413]: ok

**Review by Astralidea — DISMISSED** [review]

**Review by Seaven — DISMISSED** [review]

**Review by trueeyu — DISMISSED** [review]

**Review by Seaven — DISMISSED** [review]

**Comment by trueeyu** [comment]
@Mergifyio rebase main

**Comment by trueeyu** [comment]
https://github.com/Mergifyio rebase main

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java [hunk 2]
```diff
@@ -380,6 +382,26 @@ public static ConstantOperator dateFormat(ConstantOperator date, ConstantOperato
...
+                    returnType = VARCHAR, isMonotonic = true)
+    })
+    public static ConstantOperator jodatimeFormat(ConstantOperator date, ConstantOperator fmtLiteral) {
+        String format = fmtLiteral.getVarchar();
+        if (format.isEmpty()) {
+            return ConstantOperator.createNull(Type.VARCHAR);
+        }
+        org.joda.time.format.DateTimeFormatter formatter = DateTimeFormat.forPattern(format);
+        DateTime jodaDateTime = new DateTime(date.getDatetime()
+                .atZone(ZoneId.systemDefault()) // 将 LocalDateTime 与系统默认时区关联
```
Comment (murphyatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java:398]: please use english comment

Comment (renzhimin7) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java]: ok

**Review by murphyatwork — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by murphyatwork** [comment]
@Mergify backport branch-3.3 branch-3.2

**Pull request merged**